editor/src/components/canvas/grid-panels-state.tsx

```diff
@@ -0,0 +1,225 @@
+
+export const NumberOfColumns = 4
+export const IndexOfCanvas = 2
+
+export const GridPanelVerticalGapHalf = 6
+export const GridVerticalExtraPadding = 4
+export const GridPanelHorizontalGapHalf = 6
+export const GridHorizontalExtraPadding = 4
+
+export const ExtraHorizontalDropTargetPadding = 45
+
+export const GridPanelsNumberOfRows = 12
+
+export type Menu = 'inspector' | 'navigator'
+export type Pane = 'code-editor'
+
+export const allMenusAndPanels: Array<Menu | Pane> = [
+  'navigator',
+  'code-editor',
+  'inspector',
+  // 'preview', // Does this exist?
+]
+
+export interface GridPanelData {
+  panel: StoredPanel
+  span: number
+  index: number
+  order: number
+}
+
+export type PanelName = Menu | Pane
+
+export interface StoredPanel {
+  name: PanelName
+  type: 'menu' | 'pane'
+  uid: string
+}
+
+function storedPanel({ name, type }: { name: PanelName; type: 'menu' | 'pane' }): StoredPanel {
+  return {
+    name: name,
+    type: type,
+    uid: UUID(),
+  }
+}
+
+type StoredLayout = Array<Array<StoredPanel>>
+
+type BeforeColumn = {
+  type: 'before-column'
+  columnIndex: number
+}
+type AfterColumn = {
+  type: 'after-column'
+  columnIndex: number
+}
+type ColumnUpdate = BeforeColumn | AfterColumn
+
+type BeforeIndex = {
+  type: 'before-index'
+  columnIndex: number
+  indexInColumn: number
+}
+type AfterIndex = {
+  type: 'after-index'
+  columnIndex: number
+  indexInColumn: number
+}
+type RowUpdate = BeforeIndex | AfterIndex
+export type LayoutUpdate = ColumnUpdate | RowUpdate
+
+export const GridMenuDefaultPanels: StoredLayout = [
+  [
+    storedPanel({ name: 'navigator', type: 'menu' }),
+    storedPanel({ name: 'code-editor', type: 'pane' }),
+  ],
+  [],
+  [],
+  [storedPanel({ name: 'inspector', type: 'menu' })],
+]
+
+export function storedLayoutToResolvedPanels(stored: StoredLayout): {
+  [index in PanelName]: GridPanelData
+} {
+  const panels = accumulate<{ [index in PanelName]: GridPanelData }>({} as any, (acc) => {
+    stored.forEach((column, colIndex) => {
+      const panelsForColumn = column.length
+      column.forEach((panel, panelIndex) => {
+        acc[panel.name] = {
+          panel: panel,
+          span: GridPanelsNumberOfRows / panelsForColumn, // TODO introduce resize function
+          index: colIndex,
+          order: panelIndex,
+        }
+      })
+    })
+  })
+
+  return panels
+}
+
+/**
+ * Returns the index in the wraparound annotation, currently the values are -2, -1, 0, 1
+ */
+export function wrapAroundColIndex(index: number): number {
+  const normalized = normalizeColIndex(index)
+  if (normalized >= IndexOfCanvas) {
+    return normalized - NumberOfColumns
+  } else {
+    return normalized
+  }
+}
+
+/**
+ * Normalizes the index to 0,1,2,3
+ */
+function normalizeColIndex(index: number): number {
+  return mod(index, NumberOfColumns)
+}
+
+export function updateLayout(
+  stored: StoredLayout,
+  paneToMove: StoredPanel, // must be referentially equal to the stored panel!
```

Comment:
👀 that reliance sounds dangerous to me. Why the need for ref equality here? It looks to me like we could remove the old one purely based on the `name`, or even the `uid`

Reply:
oh that comment is outdated, will delete